objects.js

```diff
@@ -2640,87 +2640,135 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+            } else if (expressionString[0] === "(") {
+                tokens.push({ type: "lp", val: "("});
+                expressionString = expressionString.slice(1);
+            } else if (expressionString[0] === ")") {
+                tokens.push({ type: "rp", val: "("});
+                expressionString = expressionString.slice(1);
+            } else if (expressionString[0] === ",") {
+                tokens.push({ type: "comma", val: ","});
+                expressionString = expressionString.slice(1);
             }
         }
+        return tokens;
+    }
 
-        while (idx < expression.length) {
-            ch = expression[idx];
-            idx += 1;
-            switch (ch) {
-            case ' ':
-                break;
-            case '(':
-                if (inputs[operator ? 1 : 0].length) {
-                    inputs[operator ? 1 : 0] = [
-                        inputs[operator ? 1 : 0],
-                        parseInfix(nested())
-                    ];
+    function createASTFromTokens(tokens) {
+        return precedence2(tokens);
+        function precedence0(tokens) {
+            var token = tokens.shift();
+            if (token.type === "lp") {
+                var exp = createASTFromTokens(tokens);
+                if (tokens.shift().type !== "rp") {
+                    throw "Mismatched parenthesis";
+                }
+                return exp;
+            } else if (token.type === "number") {
+                return +token.val;
+            } else if (token.type === "operator") {
+                if (token.val === "-" || token.val === "+") {
+                    return [token.val, 0, precedence0(tokens)];
                 } else {
-                    inputs[operator ? 1 : 0] = parseInfix(nested());
+                    throw "Invalid prefix operator";
                 }
-                break;
-            case '-':
-            case '+':
-            case '*':
-            case '/':
-            case '%':
-                if (!operator && !inputs[0].length) {
-                    inputs[0] += ch;
-                } else if (operator) {
-                    return parseInfix(
-                        expression.slice(idx),
-                        ch,
-                        [operator, already, format(inputs[1])]
-                    );
+            } else if (token.type === "ident") {
+                if (tokens.length > 0 && tokens[0].type === "lp") {
+                    var args = [];
+                    tokens.splice(0, 1);
+                    while (tokens[0].type !== "rp") {
+                        args.push(createASTFromTokens(tokens));
+                        if (tokens[0].type === "rp") {
+                            break;
+                        }
+                        if (tokens[0].type !== "comma") {
+                            throw "No comma between arguments";
+                        }
+                    }
+                    tokens.splice(0, 1);
```

Comment:
I know, but it also returns it. Since I'm not doing anything with the result, I thought that splice was better. I'll change it if you think that shift is better though.